Retro Tapping Re-Write; Key Roll Fix

JohnRigoni opened the pull request:
## Description

The purpose of this PR is to fix the condition found with RETRO_TAPPING enabled when key rolling. The issue is that the dual function keys's retro tap keycode is not fired when a different key has been released between the dual function keys's down and up events. I am not sure if this is intended behavior or not, but does not seem to be given that retro tapping is supposed to apply if another key is not pressed. Also, none of the existing tests failed after the re-write.

I had to rewrite the logic, though the original idea is still intact. When considering key rolling from a dual function key to a dual function key, it was necessary to capture the previous modifier state and temporarily apply it to the tap code. There is a unit test for this.

This is still an early draft. I am brand new to QMK and am running the ZSA fork so I do not have the experience/ability to see this through without some assistance. I would be happy to hand this off to someone more qualified if appropriate.

⚠️ This change in current form does not consider RETRO_SHIFT nor AUTO_SHIFT_ENABLE. The ZSA fork does not seem to support RETRO_SHIFT so I am not able to test this. I also have not tested RETRO_TAPPING_PER_KEY, though that looks like it will function with no issue. It is not configured in Oryx and I have sparsely looked at keymap.c.

I ensured these changes work on my ZSA Voyager and that I could make all defaults. Thanks.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

#22916

## Checklist


- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## Media

### Before Key Roll Fix: Single to Dual Function
![single-dual_before](https://github.com/JohnRigoni/qmk_firmware/assets/38547951/c35d61b2-b747-4631-99ab-3824578c59c2)

### After Key Roll Fix: Single to Dual Function
![sigle-dual_after](https://github.com/JohnRigoni/qmk_firmware/assets/38547951/632795aa-3f7f-4114-b10e-ecd8a44eb565)

### Before Key Roll Fix: Dual to Dual Function
![dual-dual_before](https://github.com/JohnRigoni/qmk_firmware/assets/38547951/378f80b2-175e-40a4-b5ef-55ac031cd5bc)

### After Key Roll Fix: Dual to Dual Function
![dual-dual_after](https://github.com/JohnRigoni/qmk_firmware/assets/38547951/2c4dfd69-8b80-4f70-9ba4-a7c580634c7d)

JohnRigoni commented:
Should I have based this branch on develop? I did not see any conflicts or other changes so I did not think anything of it.

drashna commented:
> Should I have based this branch on develop? I did not see any conflicts or other changes so I did not think anything of it.

Yes. New features, changes to core, etc should all target develop.  And this definitely qualifies for that.   Generally, only keyboard PRs should target master... as well as important bugfixes (which this doesn't exactly quality as).

drashna commented on quantum/action.c:
```diff
@@ -47,7 +47,10 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 #if defined(RETRO_TAPPING) || defined(RETRO_TAPPING_PER_KEY) || (defined(AUTO_SHIFT_ENABLE) && defined(RETRO_SHIFT))
-int retro_tapping_counter = 0;
+bool     retro_should_tap = false;
```
I almost feel like this should be a structure, but I'm not sure that's the best case.

JohnRigoni commented:
I moreso meant that my branch was initially based off master. I did not see any differences in the PR besides my code changes so I did not create a new branch based on develop.

I have tried enabling both RETRO_TAPPING_PER_KEY and RETRO_SHIFT in the config.h that Oryx generates but have not found success. Even when my LSP is showing me the code is defined.

Also, since I intend to add one more feature, Ill set up the CI locally to catch these simple errors I keep making.

JohnRigoni commented:
It appears using `set_mod()` was leaving modifiers still active, hence last commit's test that ended in an unrelated keypress. It appears the only solution is to use oneshot modifiers, though it requires more lines of code.

Based on the Retro Shift tests at https://github.com/qmk/qmk_firmware/blob/8075003e6068d6f8fa5396972c2a17d39d14c584/tests/auto_shift/retro_shift/tap_hold_configurations/default_mod_tap/test_retro_shift.cpp#L136 and surrounding it, Retro Shift never has an instance where it is activated via keyroll. Thus, it has no need for the retro_tap mods. This also explains why there was no need to change any existing tests.

IsaacElenbaas commented:
First off, great idea for a first PR. I think having roll cases working well for people is much more effective than them doing too many tapping term shenanigans.

Stale-bot brought me back to this and I took a closer look, two things I see to address:
1. You are saving and reapplying mods. `TAPPING_TERM` should still apply to rolls with Retro Tapping, and I am not sure if this respects that. If it does, that should be tested given the current implementation method. `IGNORE_MOD_TAP_INTERRUPT` is dead (long live `HOLD_ON_OTHER_KEYPRESS`) so [this](https://web.archive.org/web/20230519045024/https://cdn.discordapp.com/attachments/451392568345755648/747388373781708930/tap-holds-qmk.pdf) is out of date, but there may be a similar doc somewhere.
2. Also with saving and reapplying mods, I believe `TAP_CODE_DELAY` should be in there? The docs state that it is only for before the release of a programmatically-tapped key, but we were doing it between every modifier and key action in auto shift even before I touched it. (People are also having issues with that not working properly which I still need to address. . . no working keyboards right now though.) @drashna thoughts?

Re: `RETRO_SHIFT`, as long as its test cases are passing don't worry about it. I looked back over them recently and I think there was one combo of key types I realized I missed, but for the most part every behavior is tested there. It is also completely separate from retro tapping aside from using the same defines and them existing around each other in action / action_tapping.

IsaacElenbaas commented:
> I have tried enabling both RETRO_TAPPING_PER_KEY and RETRO_SHIFT in the config.h that Oryx generates but have not found success. Even when my LSP is showing me the code is defined.

Again, don't worry about `RETRO_SHIFT` for the most part, but if you want to poke at it and see if both features will combine flawlessly this is probably because the keys aren't enabled on Auto Shift's side. `MT(MOD_LCTL, KC_A)` is different from `KC_A` and so will need to be [added to Auto Shift.](https://docs.qmk.fm/features/auto_shift#auto-shift-per-key) By default everything is retro shifted without `RETRO_TAPPING_PER_KEY` defined (if it is auto shifted).

JohnRigoni commented:
This should sort out both the concerns you raised.

1: This change does respect TAPPING_TERM and I have added tests for every permutation of key roll (finally got the hang of the tests).

2: I removed the use of one-shot-modifiers and registered/unregistered the mods manually. After the mods are registered and before they are unregistered, wait_ms(TAP_CODE_DELAY) is run. The code is sent via tap_code(), which has TAP_CODE_DELAY built in. That should respect the delay, and CAPSLOCK is not a modifier so it shouldn't be necessary to consider when waiting.

In regards to RETROSHIFT, since there is never an instance where it is activated by a key roll and the tests are passing, I am confident it is unaffected.

I am going to add one more test that will cover chaining 2 keyrolls.

JohnRigoni commented:
I believe this PR is ready to merge. Thanks for the help.

JohnRigoni commented:
Hi @NikGovorov, glad you are finding the changes useful. There should be no issue running both together, I have been doing so since I made this PR! I plan to submit the RETRO_TAPPING_TIMEOUT feature once this bug fix is accepted, hopefully soon.

JohnRigoni commented:
Is there anything I can do to expedite this process? Are the tests too numerous/verbose?


I would like to make another contribution to Retro Tapping that will increase its efficacy once this moves forward. Thanks.

tzarc commented:
> Is there anything I can do to expedite this process? Are the tests too numerous/verbose?
> 
> I would like to make another contribution to Retro Tapping that will increase its efficacy once this moves forward. Thanks.

Just needed another review. Thanks!

JohnRigoni commented:
Thanks @NikGovorov I'll try to make a pull request tomorrow.
